### PR TITLE
[Core] Enumerate copy so GetGesturesFor list can be mutated w\o throwing

### LINKED ISSUE #2576: Regression - Android - Exception thrown when page is popped
PopAsync after a PushAsync throws Unhandled Exception:
System.InvalidOperationException: Collection was modified; enumeration operation may not execute. occurred

Run the attached program.  It will present a list of 'Announcements'.  
1. Tap one of the items to open a second page
2. press the back button to return to the previous page
3. exception is thrown

Page will pop without exception

Exception is thrown

- Version with issue:  2.5.1.444934
- Last known good version:  2.5.0.122203
- IDE: VS2017
- Platform Target Frameworks: 
  - Android: 8.1
- Android Support Library Version: 27.0.2
- Nuget Packages:
- Affected Devices:

[PopException.zip](https://github.com/xamarin/Xamarin.Forms/files/1965013/PopException.zip)

### PR DESCRIPTION
**Pull request by kingces95**
### Description of Change ###

Enumerate copy so list can be mutated w\o throwing. Fixes #2576

### Bugs Fixed ###

fixes #2576

### API Changes ###

None

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard
- [x] Consolidate commits as makes sense

### CODE REVIEW

File: Xamarin.Forms.Core/EnumerableExtensions.cs [hunk 1]
```diff
@@ -25,6 +26,57 @@ public static class EnumerableExtensions
...
+						if (enumerator.MoveNext())
+						{
+							Count++;
+							Item2 = enumerator.Current;
+
+							if (enumerator.MoveNext())
+								Count++;
+						}
+					}
+				}
```
Comment (paymicro) [inline Xamarin.Forms.Core/EnumerableExtensions.cs]: can't we just make it easier?
```C#
Count = -1;
While (enumerator.MoveNext() && Count + 1 < Capacity)
   Items[++Count] = enumerator.Current;
```

**Comment by rmarinho** [comment]
@kingces95 did you found where we regressed ?

File: Xamarin.Forms.Core/EnumerableExtensions.cs [hunk 2]
```diff
@@ -25,22 +26,26 @@ public static class EnumerableExtensions
...
+			var result = new List<T>();
 			if (gestures == null)
-				yield break;
+				return result;
 
 			if (predicate == null)
 				predicate = x => true;
 
-			foreach (IGestureRecognizer item in gestures)
+			foreach (IGestureRecognizer item in new List<IGestureRecognizer>(gestures))
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/EnumerableExtensions.cs]: this is what fixes the issue, right ? there's no need to change the return type to List<T>. Or am I missing something ?

Comment (kingces95) [reply Xamarin.Forms.Core/EnumerableExtensions.cs]: Like 103 alone fixes the issue. @hartez asked I change the return type to list. I obliged. Now, I'm suck! :)

**Comment by StephaneDelcroix** [comment]
retargeted to 3.0.0. usually, it'd require a test or two...

**Pull request merged**